src/ngraph/op/merge.cpp

```diff
@@ -0,0 +1,67 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/op/merge.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/util.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+const string op::Merge::type_name{"Merge"};
+
+shared_ptr<Node> op::Merge::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    return make_shared<Merge>(new_args.at(0), new_args.at(1), new_args.at(2));
+}
+
+op::Merge::Merge(const Output<Node>& cond, const Output<Node>& tval, const Output<Node>& fval)
+    : Op({cond, tval, fval})
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::Merge::validate_and_infer_types()
+{
+    const PartialShape& cond_shape = get_input_partial_shape(0);
+    const PartialShape& tval_shape = get_input_partial_shape(1);
+    const PartialShape& fval_shape = get_input_partial_shape(2);
+
+    NODE_VALIDATION_CHECK(this,
+                          cond_shape.rank().is_dynamic() || 0 == size_t(cond_shape.rank()),
+                          "cond must be scalar");
+
+    NODE_VALIDATION_CHECK(this,
+                          element::boolean == get_input_element_type(0),
+                          "cond must be of type element::boolean");
+    element::Type result_et;
+
+    NODE_VALIDATION_CHECK(
+        this,
+        element::Type::merge(result_et, get_input_element_type(1), get_input_element_type(2)),
+        "Arguments do not have the same element type (tval element type: ",
+        get_input_element_type(1),
+        ", fval element type: ",
+        get_input_element_type(2),
+        ").");
+
+    NODE_VALIDATION_CHECK(
+        this, tval_shape.compatible(fval_shape), "tval and fval must have compatible shape");
+
+    PartialShape output_shape{tval_shape};
```

Comment:
I think rather than copying `tval_shape` we'll want to somehow combine info from both `tval_shape` and `fval_shape`. There are two cases I can think of, and we might pick either one:

Case 1. We require that `tval` and `fval` have the same shape. Then we can just use the existing "merge" function.

Case 2. We do _not_ require that `tval` and `fval` have the same shape. Then we would push forward a "least common denominator" shape. Can't remember if we already have a util function to compute this, but for example:

```
t:{2,3,3}, f:{2,3,3} -> out:{2,3,3}
t:{2,3,3}, f:{2,4,3} -> out:{2,?,3}
t:{2,3,3}, f:{2,?,3} -> out:{2,?,3}
t:{2,3,3}, f:{2,3} -> out:?
t:?, f:{2,3} -> out:?
```

I think it is probably the _second_ thing that we want. Option 1 winds up imposing restrictions on the runtime shape of one value (say, the t value if the cond evaluates to true) based on static information about the hypothetical runtime shape of a second value (the f value, if the cond evaluates to true) that is never actually computed, and which might not even actually exist. That seems like a weird (and in any case unnecessary) restriction.